### ISSUE TITLE
Adam optimizer with decaying learning rate

### ISSUE THREAD

**Issue by jjahanip** [tensorflow/tensorflow#18187]
I tried to implement the Adam optimizer with different beta1 and beta2 to observe the decaying learning rate changes using:
```
optimizer_obj = tf.train.optimizer(learning_rate=0.001, beta1=0.3, beta2=0.7)
```

To track the changes in learning rate, I printed the ```_lr_t``` variable of the object in the session:
```
print(sess.run(optimizer_obj._lr_t))
```

But for all iterations, I get the same value (the initialized 0.001 value).

I couldn't find how the parameters are updated in this class.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 7
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.5
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: 
```
optimizer_obj = tf.train.optimizer(learning_rate=0.001, beta1=0.3, beta2=0.7)
with tf.Session() as sess:
    print(sess.run(optimizer_obj._lr_t))
```

**Comment by sid-devic** [comment]
[This](https://stackoverflow.com/questions/37842913/tensorflow-confusion-regarding-the-adam-optimizer?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa) looks like it may address the issue you are having.

**Comment by facaiy** [comment]
Hi, I think the question is better to be asked on stackoverflow. You should use tf.minimize with your optimizer to create a train_op, and print lr_t after session.run(train_op) in each iteration.

**Comment by jjahanip** [comment]
I already checked the stackoverflow and found [this](https://stackoverflow.com/questions/37842913/tensorflow-confusion-regarding-the-adam-optimizer?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa) post explaining the concept. I am familiar with the concept. But I think there is an issue that the TensorFlow is not using the decaying learning rate. That is why I submitted this issue.

Full code to reproduce the results:

```
import tensorflow as tf

# Import MNIST data
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

def fc_layer(x, num_units, name, use_relu=True):
    in_dim = x.get_shape()[1]
    W = tf.get_variable('W_' + name,
                    dtype=tf.float32,
                    shape=[in_dim, num_units],
                    initializer=tf.truncated_normal_initializer(stddev=0.01))
    b = tf.get_variable('b_' + name,
                           dtype=tf.float32,
                           initializer=tf.constant(0., shape=[num_units], dtype=tf.float32))
    layer = tf.matmul(x, W)
    layer += b
    if use_relu:
        layer = tf.nn.relu(layer)
    return layer

# Data Dimensions
img_size_flat = 784
n_classes = 10


# Hyper-parameters
lr = 10  # The optimization initial learning rate
batch_size = 100  # Training batch size
display_freq = 100  # Frequency of displaying the training results
epochs = 10

h1 = 128  # Number of neurons in hidden layer layer.


# Create the network graph
# Placeholders for inputs (x), outputs(y)
x = tf.placeholder(tf.float32, shape=[None, img_size_flat], name='X')
y = tf.placeholder(tf.float32, shape=[None, n_classes], name='Y')

fc1 = fc_layer(x, h1, 'FC1', use_relu=True)
output_logits = fc_layer(fc1, n_classes, 'OUT', use_relu=False)

# Define the loss function, optimizer, and accuracy
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y, logits=output_logits), name='loss')

optimizer_obj = tf.train.AdamOptimizer(learning_rate=lr,beta1=0.7, beta2=.3, name='Adam-op')
optimizer = optimizer_obj.minimize(loss)

# Initialize the variables
init = tf.global_variables_initializer()


# Launch the graph (session)
with tf.Session() as sess:
    sess.run(init)

    # Number of training iterations in each epoch
    num_tr_iter = int(mnist.train.num_examples / batch_size)
    step = 0
    for epoch in range(epochs):
        for iteration in range(num_tr_iter):
            start = iteration * batch_size
            end = (iteration + 1) * batch_size
            x_batch, y_batch = mnist.train.next_batch(batch_size)

            # Run optimization op (backprop)
            feed_dict_batch = {x: x_batch, y: y_batch}
            sess.run(optimizer, feed_dict=feed_dict_batch)
            print('step:{} \t learning rate: {}'.format(step, sess.run(optimizer_obj._lr_t, feed_dict=feed_dict_batch)))
            step = step + 1
```

And this is the output that I get for each step:
```
step:0 	 learning rate: 10
step:1 	 learning rate: 10
step:2 	 learning rate: 10
step:3 	 learning rate: 10
step:4 	 learning rate: 10
step:5 	 learning rate: 10
step:6 	 learning rate: 10
step:7 	 learning rate: 10
step:8 	 learning rate: 10
step:9 	 learning rate: 10
step:10 	 learning rate: 10
step:11 	 learning rate: 10
step:12 	 learning rate: 10
step:13 	 learning rate: 10
step:14 	 learning rate: 10
step:15 	 learning rate: 10
step:16 	 learning rate: 10
step:17 	 learning rate: 10
step:18 	 learning rate: 10
step:19 	 learning rate: 10
step:20 	 learning rate: 10
step:21 	 learning rate: 10
step:22 	 learning rate: 10
step:23 	 learning rate: 10
step:24 	 learning rate: 10
step:25 	 learning rate: 10
step:26 	 learning rate: 10
step:27 	 learning rate: 10
step:28 	 learning rate: 10
step:29 	 learning rate: 10
step:30 	 learning rate: 10
step:31 	 learning rate: 10
step:32 	 learning rate: 10
step:33 	 learning rate: 10
step:34 	 learning rate: 10
step:35 	 learning rate: 10
step:36 	 learning rate: 10
step:37 	 learning rate: 10
step:38 	 learning rate: 10
step:39 	 learning rate: 10
step:40 	 learning rate: 10
step:41 	 learning rate: 10
step:42 	 learning rate: 10
step:43 	 learning rate: 10
step:44 	 learning rate: 10
step:45 	 learning rate: 10
step:46 	 learning rate: 10
step:47 	 learning rate: 10
step:48 	 learning rate: 10
step:49 	 learning rate: 10
step:50 	 learning rate: 10
```

I just output the first 50 steps. But the learning rate is fixed all the time.

**Comment by facaiy** [comment]
@jjahanip  You're right. `lr_t_` is constant, which is `learning_rate` in the [API document](https://www.tensorflow.org/versions/master/api_docs/python/tf/train/AdamOptimizer).   

https://github.com/tensorflow/tensorflow/blob/f9c5e71104cb30583127fdc918591cc7604f17ca/tensorflow/python/training/adam.py#L135

And the real decaying `lr_t` is computed as an intermediate result inside the computing function. You seems to have to compute it by yourself.
https://github.com/tensorflow/tensorflow/blob/f9c5e71104cb30583127fdc918591cc7604f17ca/tensorflow/python/training/adam.py#L176

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatatodd** [comment]
Assigning to @alextp, who probably has a better idea about this than I do.

**Comment by alextp** [comment]
@facaiy 's answer is correct, so I'm closing this issue. Please reopen with additional information if something is not clear.

**Comment by SamahZaro** [comment]
> Hi, I think the question is better to be asked on stackoverflow. You should use tf.minimize with your optimizer to create a train_op, and print lr_t after session.run(train_op) in each iteration.

Unfortunately, this doesn't work. Neither 'lr_t', '_lr_t' nor '_lr' is an attribute of 'Operation' object which returned from minimize().

I think it should be easier to have these values ready from the optimizer itself than re-computing it by explicit equations.

If anybody has an easy solution or a working code, please share it. This would be appreciated.
